// @magic-ext/webauthn/dist/commonjs/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebAuthnSDKErrorCode = exports.MagicWebAuthnPayloadMethod = void 0;
var MagicWebAuthnPayloadMethod;
(function (MagicWebAuthnPayloadMethod) {
    MagicWebAuthnPayloadMethod["WebAuthnRegistrationStart"] = "magic_auth_webauthn_registration_start";
    MagicWebAuthnPayloadMethod["RegisterWithWebAuth"] = "magic_auth_webauthn_register";
    MagicWebAuthnPayloadMethod["LoginWithWebAuthn"] = "magic_auth_login_with_web_authn";
    MagicWebAuthnPayloadMethod["WebAuthnLoginVerify"] = "magic_auth_login_with_webauthn_verify";
    MagicWebAuthnPayloadMethod["GetWebAuthnInfo"] = "magic_user_get_webauthn_credentials";
    MagicWebAuthnPayloadMethod["UpdateWebAuthnInfo"] = "magic_user_update_webauthn";
    MagicWebAuthnPayloadMethod["UnregisterWebAuthDevice"] = "magic_user_unregister_webauthn";
    MagicWebAuthnPayloadMethod["RegisterWebAuthDeviceStart"] = "magic_auth_register_webauthn_device_start";
    MagicWebAuthnPayloadMethod["RegisterWebAuthDevice"] = "magic_auth_register_webauthn_device";
})(MagicWebAuthnPayloadMethod = exports.MagicWebAuthnPayloadMethod || (exports.MagicWebAuthnPayloadMethod = {}));
var WebAuthnSDKErrorCode;
(function (WebAuthnSDKErrorCode) {
    WebAuthnSDKErrorCode["WebAuthnNotSupported"] = "WEBAUTHN_NOT_SUPPORTED";
    WebAuthnSDKErrorCode["WebAuthnCreateCredentialError"] = "WEBAUTHN_CREATE_CREDENTIAL_ERROR";
})(WebAuthnSDKErrorCode = exports.WebAuthnSDKErrorCode || (exports.WebAuthnSDKErrorCode = {}));
