// @magiclabs/ui/src/stories/1-primitives/0-CallsToAction.stories.tsx
import React from 'react';
import { action } from '@storybook/addon-actions';
import { withDesign } from 'storybook-addon-designs';
import { StoryMetadata, Story } from '../../types/storybook-metadata';
import { CallToAction } from '../../components/primitives';
import { MonochromeIcons } from '../../components/icons';

export default {
  title: 'Primitives/Call to Action',
  component: CallToAction,
  parameters: {
    docs: {
      description: {
        component:
          'A "Call to Action" button is used to prompt the user to take some specified action, such as submitting a form.',
      },
    },
  },
  decorators: [withDesign],
} as StoryMetadata;

export const CallToActionStory: Story = (args) => {
  return (
    <CallToAction leadingIcon={MonochromeIcons.Document} {...args} onPress={action('Clicked')}>
      Hello world!
    </CallToAction>
  );
};

CallToActionStory.argTypes = {
  leadingIcon: { control: { type: null } },
  trailingIcon: { control: { type: null } },
};

CallToActionStory.storyName = 'Call to Action';

CallToActionStory.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/vn0cdq2HFXKIEMyVHQNFpr/%E2%9D%96-UI-Components-v2?node-id=139%3A2831',
  },
};
