// @magiclabs/ui/src/stories/1-primitives/5-Radio.stories.tsx
import React, { useState, useCallback } from 'react';
import { withDesign } from 'storybook-addon-designs';
import { StoryMetadata, Story } from '../../types/storybook-metadata';
import { Radio, RadioGroup } from '../../components/primitives';
import { FlexColumn, Outset } from '../../components/layout';

export default {
  title: 'Primitives/Radio',
  component: Radio,
  subcomponents: { RadioGroup },
  decorators: [withDesign],
} as StoryMetadata;

export const RadioStory: Story = (args) => {
  const { checked: _checked, ...otherArgs } = args;
  const [checked, setChecked] = useState(true);
  const toggle = useCallback(() => setChecked(!checked), [checked]);

  return <Radio {...otherArgs} checked={checked} onPress={toggle} />;
};

RadioStory.argTypes = {
  checked: { control: { type: null } },
};

RadioStory.storyName = 'Radio';

RadioStory.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/vn0cdq2HFXKIEMyVHQNFpr/%E2%9D%96-UI-Components-v2?node-id=166%3A25',
  },
};

export const RadioGroupStory: Story = (args) => {
  const [selectedValue, setSelectedValue] = useState<any>();

  return (
    <RadioGroup name="foobarbaz" selectedValue={selectedValue} onChange={(e) => setSelectedValue(e.target.value)}>
      <FlexColumn>
        <Outset wrapItemsWith="div" top={5} bottom={5} trim="vertical">
          <Radio value="foo" label="Foo" />
          <Radio value="bar" label="Bar" />
          <Radio value="baz" label="Baz" />
        </Outset>
      </FlexColumn>
    </RadioGroup>
  );
};

RadioGroupStory.argTypes = {};

RadioGroupStory.storyName = 'Radio Group';

RadioGroupStory.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/vn0cdq2HFXKIEMyVHQNFpr/%E2%9D%96-UI-Components-v2?node-id=166%3A25',
  },
};
