// @magiclabs/ui/src/stories/1-primitives/6-Input.stories.tsx
import React, { useState } from 'react';
import { withDesign } from 'storybook-addon-designs';
import { StoryMetadata, Story } from '../../types/storybook-metadata';
import { Input, CopyableInput } from '../../components/primitives';
import { Icon, MonochromeIcons } from '../../components/icons';
import { Outset, FlexColumn } from '../../components/layout';

export default {
  title: 'Primitives/Input',
  component: Input,
  subcomponents: { CopyableInput },
  decorators: [withDesign],
} as StoryMetadata;

export const InputStory: Story = (args) => {
  const [errorMessage, setErrorMessage] = useState('');

  return (
    <Input
      label="Email"
      type="email"
      placeholder="hello@magic.link"
      validateOnBlur
      errorMessage={errorMessage}
      onInvalid={() => setErrorMessage('Please enter a valid email!')}
      onFocus={() => setErrorMessage('')}
      suffix={<Icon type={MonochromeIcons.Envelope} />}
      {...args}
    />
  );
};

InputStory.argTypes = {
  required: { control: { type: 'boolean' } },
  prefix: { control: { type: null } },
  suffix: { control: { type: null } },
};

InputStory.storyName = 'Input';

InputStory.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/vn0cdq2HFXKIEMyVHQNFpr/%E2%9D%96-UI-Components-v2?node-id=147%3A163',
  },
};

export const CopyableInputStory: Story = () => {
  return (
    <FlexColumn>
      <Outset wrapItemsWith="div" bottom={15} trim="vertical">
        <CopyableInput defaultValue="Copy me! (editible input)" />
        <CopyableInput defaultValue="Copy me! (disabled input)" disabled />
      </Outset>
    </FlexColumn>
  );
};

CopyableInputStory.argTypes = {
  prefix: { control: { type: null } },
  suffix: { control: { type: null } },
};

CopyableInputStory.storyName = 'Copyable Input';
