// @magiclabs/ui/src/stories/4-feedback/0-Toast.stories.tsx
import React, { useCallback } from 'react';
import { withDesign } from 'storybook-addon-designs';
import { StoryMetadata, Story } from '../../types/storybook-metadata';
import { useToast, ToastProvider } from '../../components/feedback';
import { Linkable } from '../../components/primitives';

export default {
  title: 'Feedback/Toast',
  component: ToastProvider,
  decorators: [withDesign],
} as StoryMetadata;

const ToastRender = () => {
  const { createToast } = useToast();

  const addToasts = useCallback(() => {
    createToast({ message: 'This is an info toast', type: 'info', lifespan: 4000 });
    createToast({ message: 'This is a success toast', type: 'success', lifespan: 6000 });
    createToast({ message: 'This is an error toast', type: 'error', lifespan: 2000 });
    createToast({ message: 'This is a warning toast', type: 'warning', lifespan: 8000 });
    createToast({ message: 'This toast has no icon', messageOnly: true, type: 'info', lifespan: 6000 });
  }, [createToast]);

  return (
    <Linkable>
      <button type="button" onClick={addToasts}>
        Add toasts
      </button>
    </Linkable>
  );
};

export const ToastStory: Story = (args) => {
  return (
    <ToastProvider position={typeof args.position === 'string' ? (args.position as any) : 'top'}>
      <ToastRender />
    </ToastProvider>
  );
};

ToastStory.storyName = 'Toast';

ToastStory.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/vn0cdq2HFXKIEMyVHQNFpr/%E2%9D%96-UI-Components-v2?node-id=215%3A72',
  },
};
