// @magiclabs/ui/src/stories/5-layout/0-useTabs.stories.tsx
import React from 'react';
import { withDesign } from 'storybook-addon-designs';
import { StoryMetadata, Story } from '../../types/storybook-metadata';
import { useTabs } from '../../components/layout';
import { MonochromeIcons } from '../../components/icons';

export default {
  title: 'Layout/Tabs',
  decorators: [withDesign],
} as StoryMetadata;

export const HorizontalTabsStory: Story = (args) => {
  const { tablist, tabpanel } = useTabs({
    id: 'horizontal-tabs',
    label: 'Horizontal Tabs',
    size: 'lg',

    tabs: [
      { id: 'one', label: 'First tab', content: <div>This is the first tab</div> },
      { id: 'two', label: 'Second tab with longer label', content: <div>This is the second tab</div> },
      { id: 'three', label: 'Third tab', content: <div>This is the third tab</div>, icon: MonochromeIcons.Document },
    ],
  });

  return (
    <div>
      {tablist}
      <div style={{ marginTop: '20px' }}>{tabpanel}</div>
    </div>
  );
};

HorizontalTabsStory.storyName = 'Horizontal Tabs';

HorizontalTabsStory.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/vn0cdq2HFXKIEMyVHQNFpr/%E2%9D%96-UI-Components-v2?node-id=631%3A0',
  },
};

export const VerticalTabsStory: Story = (args) => {
  const { tablist, tabpanel } = useTabs({
    id: 'vertical-tabs',
    label: 'Vertical Tabs',
    orientation: 'vertical',
    size: 'lg',

    tabs: [
      { id: 'one', label: 'First tab', content: <div>This is the first tab</div>, icon: MonochromeIcons.Envelope },
      { id: 'two', label: 'Second tab', content: <div>This is the second tab</div> },
      { id: 'three', label: 'Third tab', content: <div>This is the third tab</div> },
    ],
  });

  return (
    <div>
      {tablist}
      <div style={{ marginTop: '20px' }}>{tabpanel}</div>
    </div>
  );
};

VerticalTabsStory.storyName = 'Vertical Tabs';
