// @magiclabs/ui/dist/es/hooks/transitions.js
import { useCallback } from 'react';

function createFramerTransition() {
    return {
        withVariants: (variants) => ({
            use: (targets) => {
                const { initial, animate, exit } = targets;
                return useCallback((custom = {}) => ({ custom, variants, initial, animate, exit }), [
                    initial,
                    animate,
                    exit,
                ]);
            },
        }),
    };
}
const fadeTransition = /* @__PURE__ */ createFramerTransition().withVariants({
    visible: { opacity: 1 },
    hidden: { opacity: 0 },
});
function useFade() {
    return fadeTransition.use({
        initial: 'hidden',
        animate: 'visible',
        exit: 'hidden',
    });
}
const scaleTransition = /* @__PURE__ */ createFramerTransition().withVariants({
    hiddenWithScale: (scale) => ({ opacity: 0, scale, transition: { duration: 0.1 } }),
    hiddenWithoutScale: { opacity: 0, scale: 1, transition: { duration: 0.1 } },
    visible: { opacity: 1, scale: 1, transition: { duration: 0.1 } },
});
function useScale() {
    return scaleTransition.use({
        initial: 'hiddenWithScale',
        animate: 'visible',
        exit: 'hiddenWithScale',
    });
}
function useScaleIn() {
    return scaleTransition.use({
        initial: 'hiddenWithScale',
        animate: 'visible',
        exit: 'hiddenWithoutScale',
    });
}
function useScaleOut() {
    return scaleTransition.use({
        initial: 'hiddenWithoutScale',
        animate: 'visible',
        exit: 'hiddenWithScale',
    });
}
const slideTransition = /* @__PURE__ */ createFramerTransition().withVariants({
    hidden: ({ x = 0, y = 0 }) => ({ x, y, opacity: 0, transition: { ease: 'easeInOut', duration: 0.15 } }),
    visible: { x: 0, y: 0, opacity: 1, transition: { ease: 'easeInOut', duration: 0.15 } },
});
function useSlide() {
    return slideTransition.use({
        initial: 'hidden',
        animate: 'visible',
        exit: 'hidden',
    });
}

export { createFramerTransition, useFade, useScale, useScaleIn, useScaleOut, useSlide };
